File: src/0740_delete_and_earn.js
function deleteAndEarn(nums) {
    const upperLimit = Math.max(...nums) + 1;
    const store = Array(upperLimit).fill(0);
    for (const num of nums) {
        store[num] += num;
    }
    const dp = Array(upperLimit).fill(0);
    if (upperLimit > 1) {
        dp[1] = store[1];
    }
    for (let i = 2; i < upperLimit; i++) {
        dp[i] = Math.max(dp[i - 2] + store[i], dp[i - 1]);
    }
    return dp[upperLimit - 1];
}

module.exports = { deleteAndEarn };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(deleteAndEarn([3, 4, 2]), 6);
    assert.deepStrictEqual(deleteAndEarn([2, 2, 3, 3, 3, 4]), 9);
}

File: src/0740_delete_and_earn.rs
pub fn delete_and_earn(nums: Vec<i32>) -> i32 {
    let upper_limit = nums.iter().copied().max().unwrap_or(0) as usize + 1;
    let mut store = vec![0; upper_limit];

    for num in nums {
        store[num as usize] += num;
    }

    let mut dp = vec![0; upper_limit];
    if upper_limit > 1 {
        dp[1] = store[1];
    }
    for i in 2..upper_limit {
        dp[i] = (dp[i - 2] + store[i]).max(dp[i - 1]);
    }

    dp[upper_limit - 1]
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mirrors_python_main_assertions() {
        assert_eq!(delete_and_earn(vec![3, 4, 2]), 6);
        assert_eq!(delete_and_earn(vec![2, 2, 3, 3, 3, 4]), 9);
    }
}
